astropy/coordinates/builtin_frames/icrs_cirs_transforms.py

```diff
@@ -284,30 +284,48 @@ def gcrs_to_hcrs(gcrs_coo, hcrs_frame):
                      "function in this case because there is an origin shift.")
 
 
-@frame_transform_graph.transform(FunctionTransform, HCRS, ICRS)
+@frame_transform_graph.transform(AffineTransform, HCRS, ICRS)
 def hcrs_to_icrs(hcrs_coo, icrs_frame):
     # this is just an origin translation so without a distance it cannot go ahead
     if hcrs_coo.data.__class__ == UnitSphericalRepresentation:
         raise u.UnitsError(_NEED_ORIGIN_HINT.format(hcrs_coo.__class__.__name__))
 
-    # this goes here to avoid circular import errors
-    from ..solar_system import get_body_barycentric
-    bary_sun_pos = get_body_barycentric('sun', hcrs_coo.obstime)
-    newrep = hcrs_coo.cartesian + bary_sun_pos
-    return icrs_frame.realize_frame(newrep)
+    bary_sun_vel = None
+    if hcrs_coo.data.differentials:
+        from ..solar_system import get_body_barycentric_posvel
+        bary_sun_pos, bary_sun_vel = get_body_barycentric_posvel('sun',
+                                                                 hcrs_coo.obstime)
+        dv = bary_sun_vel.xyz
+
+    else:
+        from ..solar_system import get_body_barycentric
+        bary_sun_pos = get_body_barycentric('sun', hcrs_coo.obstime)
+        dv = None
 
+    dx = bary_sun_pos.xyz
 
-@frame_transform_graph.transform(FunctionTransform, ICRS, HCRS)
+    return None, (dx, dv)
+
+@frame_transform_graph.transform(AffineTransform, ICRS, HCRS)
 def icrs_to_hcrs(icrs_coo, hcrs_frame):
     # this is just an origin translation so without a distance it cannot go ahead
     if icrs_coo.data.__class__ == UnitSphericalRepresentation:
```

Comment:
While you are here, replace `==` with `is` (or do `isinstance(icrs_coo.data, UnitSphericalRepresentation)`)